back-end/utils/top-layer/UserUtil.js

```diff
@@ -362,6 +362,173 @@ const getLengthUserTransactionsHistoryForRedisM5RU = (email, filters) => {
   });
 };
 
+/**
+ * Perform buy action for the function proceedTransaction
+ * @param {import(".prisma/client").User} user
+ * @param {number} totalPendingPrice
+ * @param {string} companyCode
+ * @returns
+ */
+ const buyShareEvent = async (
+  user,
+  totalPendingPrice,
+  companyCode,
+  quantity,
+  recentPrice
+) => {
+  try {
+    if (user.cash < totalPendingPrice) {
+      return new Error(
+        "The user does not have enough money to buy the pending stock"
+      );
+    }
+
+    user.cash -= totalPendingPrice;
+    const buyShare = user.shares.find(
+      (share) => share.companyCode === companyCode
+    );
+    const buyShareIndex = user.shares.indexOf(buyShare);
+
+    if (!buyShare) {
+      await prisma.share.create({
+        companyCode: companyCode,
+        quantity: quantity,
+        buyPriceAvg: recentPrice,
+        user: user,
+        userID: user.id
+      });
+    } else {
+      // Update in Share
+      buyShare.buyPriceAvg =
```

Comment:
This formula might be incorrect. The correct formula might be:
```
(buyShare.buyPriceAvg * buyShare.quantity +recentPrice * quantity) / (quantity + buyShare.quantity)
```
Consider adding comment to explain the formula.